keps/sig-api-machinery/3920-finish-feature-flags/README.md

```diff
@@ -0,0 +1,973 @@
+transition. Since this is a distributed system and components can fail, this
+list needs to be stored in the Feature.
+
+What is a task and why do we need them?? Examples:
+* Waiting for all uses of the feature to match the configured state. This needs
+  to happen for every transition, so we will implement this -- no work needed
+  from feature authors.
+* Touch all objects in etcd to remove a field whose feature has been turned off.
+  Only needed by features which are adding a field to some API object. We will
+  have a place in the code to announce the need for the task, declare a function
+  to perform the task, and we will run that function when needed. But the
+  feature author will need to write the function to perform this operation.
+* Suppose the feature is a new volume type; then a task might be to wait for
+  volumes of that type to be removed for the system. It might be dangerous to
+  automatically remove them. While they exist, we need the feature to stay in
+  the TurningOff state.
+
+Completing state transitions. When we evaluate a feature that is in a
+transitional state and has tasks remaining, we attempt to perform the first
+task. If it is successful, we remove it from the list. If that was the last
+task, then we also change the feature state to the final state, On or Off.
+
+Waiting for uses. This task examines the uses the client flow has written to the
+object. Specifically, we clear the uses, wait a specified amount of time, and
+see if any client checks in with the old value of the feature or not. Once no
+client reports the old value, this task is complete.
+
+### User Stories
+
+Since the primary concerns with this proposal have been around whether it can
+support existing features without requiring too much work from feature authors,
+we will use this section to explain how several different features will work.
+
+#### Story: MaxUnavailableStatefulSet
+
+This feature adds a field to StatefulSet and logic in the controller.
+
+When this feature is turned off, we would like to use the server flow logic
+described above to remove any use of the field from api objects.
+
+Additionally this feature could benefit from only needing to be turned on in one
+place (apiserver command line) instead of also needing to be set on
+controler-manager. (Secondary goals)
+
+Additionally we should remove the field from the published schema if the feature
+is off (follow-up goals).
+
+#### Story: OpenAPIV3
+
+This feature enables serving OpenAPI v3 spec documents in apiserver. It is
+stateless and there is no work needed from the server flow.
+
+It also affects only one binary (apiserver).
+
+It is GA now, so this isn't a concern, but early in their lifecycle, low-impact
+features like this would benefit from being dynamically settable, so cluster
+admins without access to apiserver's command line could turn it on. (Secondary
+goals)
+
+#### Story: Feature which takes additional config on command line
```

Comment:
For kubelet, NodeSwap in kubelet needs `failSwapOn=false`.  I cannot find one in apiserver.

Reply:
thank you! I will add this.

Reply:
Added.